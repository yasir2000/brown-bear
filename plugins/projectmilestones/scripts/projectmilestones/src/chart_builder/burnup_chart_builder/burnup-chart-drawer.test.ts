/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { createBurnupChart, getTotal } from "./burnup-chart-drawer";
import type { ChartPropsWhithoutTooltip } from "../../../../../../../src/scripts/charts-builders/type";
import type {
    GenericBurnupData,
    PointsWithDateForGenericBurnup,
} from "../../../../../../agiledashboard/scripts/burnup-chart/src/type";
jest.mock("../../../../../../../src/scripts/charts-builders/time-scale-labels-formatter");
jest.mock("../time-scale-label-formatter");

describe("BurnupChartDrawer", () => {
    describe("getTotal", () => {
        it("When there is a highest total in points, Then it returns", () => {
            const generic: GenericBurnupData = {
                capacity: null,
                points_with_date: [
                    {
                        date: new Date().toDateString(),
                        progression: 10,
                        total: 15,
                    },
                    {
                        date: new Date().toDateString(),
                        progression: 10,
                        total: 20,
                    },
                ],
            } as GenericBurnupData;

            const total = getTotal(generic);
            expect(total).toEqual(20);
        });

        it("When points are null, Then capacity returns", () => {
            const generic: GenericBurnupData = {
                capacity: 30,
                points_with_date: [] as PointsWithDateForGenericBurnup[],
            } as GenericBurnupData;

            const total = getTotal(generic);
            expect(total).toEqual(30);
        });

        it("When points and capacity are null, Then DEFAULT_TOTAL_EFFORT returns", () => {
            const generic: GenericBurnupData = {
                capacity: null,
                points_with_date: [] as PointsWithDateForGenericBurnup[],
            } as GenericBurnupData;

            const total = getTotal(generic);
            expect(total).toEqual(5);
        });
    });
    describe("createBurnupChart", () => {
        it("When the chart is created, Then there are a G element and 2 lines scale with scale label", () => {
            const chart_svg_element = getDocument();
            createBurnupChart(chart_svg_element, getChartProps(), getGenericBurnupData());

            expect(chart_svg_element).toMatchSnapshot();
        });

        function getDocument(): HTMLElement {
            const local_document = document.implementation.createHTMLDocument();
            const chart_div = local_document.createElement("svg");
            chart_div.setAttribute("id", "chart-100");
            return chart_div;
        }

        function getChartProps(): ChartPropsWhithoutTooltip {
            return {
                graph_width: 100,
                graph_height: 100,
                margins: {
                    top: 20,
                    right: 20,
                    bottom: 20,
                    left: 20,
                },
            };
        }

        function getGenericBurnupData(): GenericBurnupData {
            return {
                opening_days: [1, 2, 3, 4, 5],
                duration: 1,
                start_date: "2019-07-01T00:00:00+00:00",
                capacity: null,
                is_under_calculation: false,
                points_with_date: getGenericMaxIs15(),
            };
        }

        function getGenericMaxIs15(): PointsWithDateForGenericBurnup[] {
            const points: PointsWithDateForGenericBurnup[] = [];
            points.push({
                date: "2019-07-01T00:00:00+00:00",
                progression: 10,
                total: 15,
            });
            points.push({
                date: "2019-07-01T00:00:00+00:00",
                progression: 12,
                total: 15,
            });

            return points;
        }
    });
});
