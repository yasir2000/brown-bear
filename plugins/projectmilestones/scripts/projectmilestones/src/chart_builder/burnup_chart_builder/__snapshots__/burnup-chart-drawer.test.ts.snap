// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BurnupChartDrawer createBurnupChart When the chart is created, Then there are a G element and 2 lines scale with scale label 1`] = `
<svg
  id="chart-100"
>
  <svg
    height="100"
    width="100"
  >
    <g
      class="chart-x-axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, 80)"
    >
      <g
        class="tick"
        opacity="1"
        transform="translate(20.5,0)"
      >
        <text
          dy="0.71em"
          fill="currentColor"
          y="11"
        >
          2019-07-01
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(80.5,0)"
      >
        <text
          dy="0.71em"
          fill="currentColor"
          y="11"
        >
          2019-07-02
        </text>
      </g>
    </g>
    <g
      class="chart-y-axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(20, 0)"
    >
      <g
        class="tick"
        opacity="1"
        transform="translate(0,80.5)"
      >
        <text
          dy="0.32em"
          fill="currentColor"
          x="-5"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,61.75)"
      >
        <text
          dy="0.32em"
          fill="currentColor"
          x="-5"
        >
          5
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,43)"
      >
        <text
          dy="0.32em"
          fill="currentColor"
          x="-5"
        >
          10
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,24.25)"
      >
        <text
          dy="0.32em"
          fill="currentColor"
          x="-5"
        >
          15
        </text>
      </g>
    </g>
    <g
      class="ideal-line"
    >
      <path
        d="M20,80L80,23.75"
      />
    </g>
    <line
      class="release-line-scale"
      x1="20"
      x2="90"
      y1="80"
      y2="80"
    />
    <line
      class="release-line-scale"
      x1="20"
      x2="20"
      y1="80"
      y2="13.75"
    />
    <path
      class="chart-curve-total"
      d="M20,23.75Z"
    />
    <path
      class="chart-curve-progression"
      d="M20,42.5L20,35"
    />
  </svg>
</svg>
`;
